### ISSUE TITLE
[EXPO51][metro] bundling fails Error: Unexpected token 'v', "var __BUND"... is not valid JSON (react native's config) or Unable to resolve (expo's config)

### ISSUE THREAD

**Issue by taekeun-two** [expo/expo#30886]
### Minimal reproducible example

I couldn't share the minimal reproducible example repo, but in similar problems, other guys share it.

### What platform(s) does this occur on?

Android, iOS

### Where did you reproduce the issue?

in a development build

## Background

Recently upgraded from react-native `0.72.10` to latest expo 51 and react-native `0.74.5`

    * bare workflow having `ios` and `android` folder
    "expo": "~51.0.25",
    "react-native": "0.74.5",

* Create new project using latest command `npx create-expo-app`
* Check and add used libraries and apply corresponding native change if necessary.
* Apply legacy source code
* Check build

Personally I have no idea on metro and experience on bundling and just shared what I have done to solve this problem.

## Similar problems

As I searched, it seems that the following problems are similar one.
I couldn't share the minimal reproducible example repo, but the below link has it.

* https://stackoverflow.com/questions/78762576/metro-error-serializer-did-not-return-expected-format-error-unexpected-to
* https://www.reddit.com/r/expo/comments/1e5yhhy/metro_error_serializer_did_not_return_expected/

## [1] : When EXPO `metro.config.js` used, bundling fails.

When expo's `metro.config.js` is used, the bundling fails.

https://github.com/kristerkari/react-native-svg-transformer?tab=readme-ov-file#step-3-configure-the-react-native-packager
```javascript
const { getDefaultConfig } = require("expo/metro-config");

module.exports = (() => {
  const config = getDefaultConfig(__dirname);

  const { transformer, resolver } = config;

  config.transformer = {
    ...transformer,
    babelTransformerPath: require.resolve("react-native-svg-transformer/expo")
  };
  config.resolver = {
    ...resolver,
    assetExts: resolver.assetExts.filter((ext) => ext !== "svg"),
    sourceExts: [...resolver.sourceExts, "svg"]
  };

  return config;
})();
```

```
Android Bundling failed 17044ms index.js (3376 modules)
Unable to resolve "./app-banner.db" from "node_modules/...SOMEWHERE_OUR_CODE.../index.js"
```

## [2] : When react-native `metro.config.js`, local build works, but EAS BUILD fails.

Because expo's `metro.config.js` fails, tried to use react-native `metro.config.js`.

After a lot of tries, finally I found that it works with the following metro configuration.

https://github.com/kristerkari/react-native-svg-transformer?tab=readme-ov-file#for-react-native-v0721-or-newer

```javascript
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const config = {
  transformer: {
    assetPlugins: ['expo-asset/tools/hashAssetFiles'],
    babelTransformerPath: require.resolve('react-native-svg-transformer/react-native'),
  },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== 'svg'),
    sourceExts: [...sourceExts, 'svg'],
  },
};

module.exports = mergeConfig(defaultConfig, config);
```

With this configuration, local development works, but EAS BUILD fails in both Android and iOS.

```
[RUN_GRADLEW] Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
[RUN_GRADLEW] Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
```

```
[RUN_GRADLEW] Starting Metro Bundler
[RUN_GRADLEW] warning: Bundler cache is empty, rebuilding (this may take a minute)
[RUN_GRADLEW] Android Bundled 22363ms index.js (3401 modules)
[RUN_GRADLEW] Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
[RUN_GRADLEW] Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
[RUN_GRADLEW]     at MetroBundlerDevServer._bundleDirectAsync (/private/var/folders/dz/qvklwtvn3pd4kd27k_0k16sr0000gn/T/eas-build-local-nodejs/4add8d80-943b-443b-af8f-25ee530c599b/build/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:1104:17)
[RUN_GRADLEW] at processTicksAndRejections (node:internal/process/task_queues:95:5)
[RUN_GRADLEW]     at MetroBundlerDevServer.metroLoadModuleContents (/private/var/folders/dz/qvklwtvn3pd4kd27k_0k16sr0000gn/T/eas-build-local-nodejs/4add8d80-943b-443b-af8f-25ee530c599b/build/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:442:21)
[RUN_GRADLEW]     at MetroBundlerDevServer.legacySinglePageExportBundleAsync (/private/var/folders/dz/qvklwtvn3pd4kd27k_0k16sr0000gn/T/eas-build-local-nodejs/4add8d80-943b-443b-af8f-25ee530c599b/build/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:553:20)
[RUN_GRADLEW]     at exportEmbedBundleAndAssetsAsync (/private/var/folders/dz/qvklwtvn3pd4kd27k_0k16sr0000gn/T/eas-build-local-nodejs/4add8d80-943b-443b-af8f-25ee530c599b/build/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:116:21)
[RUN_GRADLEW]     at exportEmbedAsync (/private/var/folders/dz/qvklwtvn3pd4kd27k_0k16sr0000gn/T/eas-build-local-nodejs/4add8d80-943b-443b-af8f-25ee530c599b/build/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:68:30)
[RUN_GRADLEW] > Task :app:createBundleReleaseJsAndAssets FAILED
[RUN_GRADLEW] FAILURE:
[RUN_GRADLEW] Build failed with an exception.
[RUN_GRADLEW] * What went wrong:
[RUN_GRADLEW] Execution failed for task ':app:createBundleReleaseJsAndAssets'.
```

## [3] With react-native `metro.config.js` and remove expo-cli bundling in native layer, it works in local/EAS BUILD, but EAS UPDATE fails

I thought that there are something wrong with expo's bundling because it works after changing to use react-native metro configuration.

I removed the newly added expo bundling handling in native layer according to the link
https://docs.expo.dev/versions/latest/config/metro/#bare-workflow-setup

android
https://docs.expo.dev/versions/latest/config/metro/#androidappbuildgradle

```
react {
...
    // Use Expo CLI to bundle the app, this ensures the Metro config
    // works correctly with Expo projects.
    // cliFile = new File(["node", "--print", "require.resolve('@expo/cli', { paths: [require.resolve('expo/package.json')] })"].execute(null, rootDir).text.trim())
    // bundleCommand = "export:embed"
...
```

iOS
https://docs.expo.dev/versions/latest/config/metro/#bundle-react-native-code-and-images
```
/* Begin PBXShellScriptBuildPhase section */
		00DD1BFF1BD5951E006B06BC /* Bundle React Native code and images */ = {

// remove expo bundling
			shellScript = "if [[ -f \"$PODS_ROOT/../.xcode.env\" ]]; then\n  source \"$PODS_ROOT/../.xcode.env\"\nfi\nif [[ -f \"$PODS_ROOT/../.xcode.env.local\" ]]; then\n  source \"$PODS_ROOT/../.xcode.env.local\"\nfi\n\n# The project root by default is one level up from the ios directory\nexport PROJECT_ROOT=\"$PROJECT_DIR\"/..\n\nif [[ \"$CONFIGURATION\" = *Debug* ]]; then\n  export SKIP_BUNDLING=1\nfi\nif [[ -z \"$ENTRY_FILE\" ]]; then\n  # Set the entry JS file using the bundler's entry resolution.\n  export ENTRY_FILE=\"$(\"$NODE_BINARY\" -e \"require('expo/scripts/resolveAppEntry')\" \"$PROJECT_ROOT\" ios absolute | tail -n 1)\"\nfi\n\nif [[ -z \"$CLI_PATH\" ]]; then\n  # Use Expo CLI\n  export CLI_PATH=\"$(\"$NODE_BINARY\" --print \"require.resolve('@expo/cli', { paths: [require.resolve('expo/package.json')] })\")\"\nfi\nif [[ -z \"$BUNDLE_COMMAND\" ]]; then\n  # Default Expo CLI command for bundling\n  export BUNDLE_COMMAND=\"export:embed\"\nfi\n\n# Source .xcode.env.updates if it exists to allow\n# SKIP_BUNDLING to be unset if needed\nif [[ -f \"$PODS_ROOT/../.xcode.env.updates\" ]]; then\n  source \"$PODS_ROOT/../.xcode.env.updates\"\nfi\n# Source local changes to allow overrides\n# if needed\nif [[ -f \"$PODS_ROOT/../.xcode.env.local\" ]]; then\n  source \"$PODS_ROOT/../.xcode.env.local\"\nfi\n\n`\"$NODE_BINARY\" --print \"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\"`\n\n";
		};

// replace legacy react-natie bundling
			shellScript = "if [[ -f \"$PODS_ROOT/../.xcode.env\" ]]; then\n  source \"$PODS_ROOT/../.xcode.env\"\nfi\nif [[ -f \"$PODS_ROOT/../.xcode.env.local\" ]]; then\n  source \"$PODS_ROOT/../.xcode.env.local\"\nfi\n\n# The project root by default is one level up from the ios directory\nexport PROJECT_ROOT=\"$PROJECT_DIR\"/..\n\nif [[ \"$CONFIGURATION\" = *Debug* ]]; then\n  export SKIP_BUNDLING=1\nfi\nif [[ -z \"$ENTRY_FILE\" ]]; then\n  # Set the entry JS file using the bundler's entry resolution.\n  export ENTRY_FILE=\"$(\"$NODE_BINARY\" -e \"require('expo/scripts/resolveAppEntry')\" $PROJECT_ROOT ios relative | tail -n 1)\"\nfi\n\n`\"$NODE_BINARY\" --print \"require('path').dirname(require.resolve('react-native/package.json')) + '/scripts/react-native-xcode.sh'\"`\n\n";
		};
```

With these changes (use react native bundling), it works in both local and EAS BUILD.
Finally I tried EAS UPDATE, it fails and it shows the exactly same error message with [2]

```base
❯ eas update --branch preview --message "[eas] check eas-update"
[expo-cli] Starting Metro Bundler
[expo-cli]
[expo-cli] Android Bundled 12218ms index.js (3387 modules)
[expo-cli] Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
[expo-cli] Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
[expo-cli]     at MetroBundlerDevServer._bundleDirectAsync (/REPO/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:1104:17)
[expo-cli]     at processTicksAndRejections (node:internal/process/task_queues:95:5)
[expo-cli]     at MetroBundlerDevServer.metroLoadModuleContents (/REPO/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:442:21)
[expo-cli]     at MetroBundlerDevServer.legacySinglePageExportBundleAsync (/REPO/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:553:20)
[expo-cli]     at /REPO/node_modules/@expo/cli/src/export/exportApp.ts:137:26
[expo-cli]     at async Promise.all (index 1)
[expo-cli]     at exportAppAsync (/REPO/node_modules/@expo/cli/src/export/exportApp.ts:127:7)
[expo-cli]     at exportAsync (/REPO/node_modules/@expo/cli/src/export/exportAsync.ts:19:3)
✖ Export failed
/REPO/node_modules/expo/bin/cli exited with non-zero code: 1
    Error: update command failed.
```

### Environment

❯ npx expo-env-info
Need to install the following packages:
expo-env-info@1.2.0
Ok to proceed? (y)

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.9.0 - ~/.nvm/versions/node/v20.9.0/bin/node
      Yarn: 1.22.22 - ~/.nvm/versions/node/v20.9.0/bin/yarn
      npm: 10.1.0 - ~/.nvm/versions/node/v20.9.0/bin/npm
      Watchman: 2024.03.25.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
    IDEs:
      Android Studio: 2024.1 AI-241.18034.62.2411.12071903
      Xcode: 15.3/15E204a - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.25 => 51.0.25
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.74.5 => 0.74.5
    npmGlobalPackages:
      eas-cli: 10.2.2
      expo-cli: 6.3.10
    Expo Workflow: bare

### Expo Doctor Diagnostics

❯ npx expo-doctor@latest
Need to install the following packages:
expo-doctor@1.9.0
Ok to proceed? (y)
✔ Check Expo config for common issues
✖ Check package.json for common issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check native tooling versions
✔ Check that packages match versions required by installed Expo SDK
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK

Detailed check results:

The following scripts in package.json conflict with the contents of node_modules/.bin: tsc.

**Comment by baryon** [comment]
"expo": "51.0.15" is the last version without this issue.

**Comment by taekeun-two** [comment]
I tried expo 51.0.15 version.

* uses bare workflow
* First setup expo recommended latest versions : expo(51.0.26), react-native(0.74.5)
* Install only mentioned expo (51.0.15) version

## Use expo's metro config like [1]

It shows the same error as before...

```
Android Bundling failed 17044ms index.js (3376 modules)
Unable to resolve "./app-banner.db" from "node_modules/...SOMEWHERE_OUR_CODE.../index.js"
```

[Q] Should this expo metro config be used only for managed workflow ?

## Use react native's metro config like [2] without native change [3]

* Use native code without change in [2]
* Just use react native's metro config without changing native build script change in [3]

```javascript
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const config = {
  transformer: {
    assetPlugins: ['expo-asset/tools/hashAssetFiles'],
    babelTransformerPath: require.resolve('react-native-svg-transformer/react-native'),
  },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== 'svg'),
    sourceExts: [...sourceExts, 'svg'],
  },
};

module.exports = mergeConfig(defaultConfig, config);
```

It works in both local and EAS BUILD.

Finally I tried EAS UPDATE. I could use `react-native`'s metro config in latest expo(51.0.26), but this version expo(51.0.15) only permit expo's metro config work.

```bash
❯ npx eas update --channel preview --message "[eas] check eas-update"
[expo-cli] CommandError: The Metro bundler configuration is missing required features from 'expo/metro-config' and cannot be used with Expo CLI. Ensure the metro.config.js file is extending 'expo/metro-config'. Learn more: https://docs.expo.dev/guides/customizing-metro
✖ Export failed
/REPO/node_modules/expo/bin/cli exited with non-zero code: 1
    Error: update command failed.
```

## Summary

### expo(51.0.26) 
* EAS-BUILD
    * expo's metro config: fails 
    ```
    Unable to resolve "./app-banner.db" from "node_modules/...SOMEWHERE_OUR_CODE.../index.js"
    ```
    * react-native's metro config: OK
* EAS-UPDATE 
    * react-native's metro config can be used.
    * error occurs
    ```
    Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON
    ```

### expo(51.0.15)

* EAS-BUILD 
    * expo's metro config: fails
    ```
    Unable to resolve "./app-banner.db" from "node_modules/...SOMEWHERE_OUR_CODE.../index.js"
    ```
    * react-native's metro config: OK
* EAS-UPDATE
    * react-native's metro config cannot be used.
```bash
❯ npx eas update --channel preview --message "[eas] check eas-update"
[expo-cli] CommandError: The Metro bundler configuration is missing required features from 'expo/metro-config' and cannot be used with Expo CLI. Ensure the metro.config.js file is extending 'expo/metro-config'. Learn more: https://docs.expo.dev/guides/customizing-metro
✖ Export failed
/REPO/node_modules/expo/bin/cli exited with non-zero code: 1
    Error: update command failed.
```

**Comment by StefanWallin** [comment]
I'm having the same issue, only on Android, not iOS. Both building on my Mac M1 and on Github Actions ubuntu-latest

- Mode: `bare`
- React Native: `0.74.5`
- expo: `51.0.26`
- expo-barcode-scanner: `13.0.1`
- expo-camera: `15.0.14`
- expo-constants: `16.0.2`
- expo-linear-gradient: `13.0.2`

### Ran: `cd android; ./gradlew clean build`
```
> Task :app:createBundleReleaseJsAndAssets
Android Bundled 67141ms index.js (2373 modules)
Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token v in JSON at position 0
Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token v in JSON at position 0
    at MetroBundlerDevServer._bundleDirectAsync (/Users/festiz/iteam/milepost_app/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:1104:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at MetroBundlerDevServer.metroLoadModuleContents (/Users/festiz/iteam/milepost_app/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:442:21)
    at MetroBundlerDevServer.legacySinglePageExportBundleAsync (/Users/festiz/iteam/milepost_app/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:553:20)
    at exportEmbedBundleAndAssetsAsync (/Users/festiz/iteam/milepost_app/app/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:116:21)
    at exportEmbedAsync (/Users/festiz/iteam/milepost_app/app/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:68:30)

> Task :app:createBundleReleaseJsAndAssets FAILED
```

### My `metro.config.js`
```javascript
/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config')

const config = {}

module.exports = mergeConfig(getDefaultConfig(__dirname), config)

```

**Comment by StefanWallin** [comment]
I'm having the same issue, only on Android, not iOS. Both building on my Mac M1 and on Github Actions ubuntu-latest

- Mode: `bare`
- React Native: `0.74.5`
- expo: `51.0.26`
- expo-barcode-scanner: `13.0.1`
- expo-camera: `15.0.14`
- expo-constants: `16.0.2`
- expo-linear-gradient: `13.0.2`

### Ran: `cd android; ./gradlew clean build`
```
> Task :app:createBundleReleaseJsAndAssets
Android Bundled 67141ms index.js (2373 modules)
Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token v in JSON at position 0
Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token v in JSON at position 0
    at MetroBundlerDevServer._bundleDirectAsync (/Users/festiz/iteam/milepost_app/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:1104:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at MetroBundlerDevServer.metroLoadModuleContents (/Users/me/appproject/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:442:21)
    at MetroBundlerDevServer.legacySinglePageExportBundleAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:553:20)
    at exportEmbedBundleAndAssetsAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:116:21)
    at exportEmbedAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:68:30)

> Task :app:createBundleReleaseJsAndAssets FAILED
```

### My `metro.config.js`
```javascript
/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config')

const config = {}

module.exports = mergeConfig(getDefaultConfig(__dirname), config)

```

**Comment by StefanWallin** [comment]
I'm having the same issue, only on Android, not iOS. Both building on my Mac M1 and on Github Actions ubuntu-latest

- Mode: `bare`
- React Native: `0.74.5`
- expo: `51.0.26`
- expo-barcode-scanner: `13.0.1`
- expo-camera: `15.0.14`
- expo-constants: `16.0.2`
- expo-linear-gradient: `13.0.2`

### Ran: `cd android; ./gradlew clean build`
```
> Task :app:createBundleReleaseJsAndAssets
Android Bundled 67141ms index.js (2373 modules)
Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token v in JSON at position 0
Error: Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token v in JSON at position 0
    at MetroBundlerDevServer._bundleDirectAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:1104:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at MetroBundlerDevServer.metroLoadModuleContents (/Users/me/appproject/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:442:21)
    at MetroBundlerDevServer.legacySinglePageExportBundleAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:553:20)
    at exportEmbedBundleAndAssetsAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:116:21)
    at exportEmbedAsync (/Users/me/appproject/app/node_modules/@expo/cli/src/export/embed/exportEmbedAsync.ts:68:30)

> Task :app:createBundleReleaseJsAndAssets FAILED
```

### My `metro.config.js`
```javascript
/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config')

const config = {}

module.exports = mergeConfig(getDefaultConfig(__dirname), config)

```

**Comment by brentvatne** [comment]
hi there, you should make sure you extend the expo metro config:

https://github.com/expo/expo/blob/main/templates/expo-template-bare-minimum/metro.config.js

```
// Learn more https://docs.expo.io/guides/customizing-metro
const { getDefaultConfig } = require('expo/metro-config');

/** @type {import('expo/metro-config').MetroConfig} */
const config = getDefaultConfig(__dirname);

module.exports = config;
```


https://docs.expo.dev/guides/customizing-metro/

**Comment by taekeun-two** [comment]
@brentvatne As I wrote, the reason that I use react native's metro config is becuase it fails when expo's metro config is used.

Is there something wrong with my project configuration ?

```
Unable to resolve "./SOMEWHERE_OUR_CODE" from "node_modules/...SOMEWHERE_OUR_CODE.../index.js"
```

**Comment by RushitaRn25** [comment]
Showing Recent Messages
Serializer did not return expected format. The project copy of `expo/metro-config` may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON

**Comment by taekeun-two** [comment]
@RushitaRn25 I also think so and it needs the help of expo team. 
But this issue was closed by @brentvatne.

**Comment by RushitaRn25** [comment]
Showing Recent Messages
Serializer did not return expected format. The project copy of expo/metro-config may be out of date. Error: Unexpected token 'v', "var __BUND"... is not valid JSON

This issue i am facing while archive build app is work good while debug mode 

"react-native": "0.74.3",
"expo": "^51.0.15",

**Comment by taekeun-two** [comment]
@RushitaRn25 Thanks for replying. @baryon  also mentioned expo `51.0.15` version is the latest working version.
I tested expo `51.0.15` with react-native `0.74.5`  and it failed at that time.
https://github.com/expo/expo/issues/30886#issuecomment-2276991155

I will try again with these versions

```
"react-native": "0.74.3",
"expo": "^51.0.15",
```

**Comment by taekeun-two** [comment]
Finally I managed to make my code work and want to share what I've done.

## TL; DR

* For latest, `expo`'s metro config should be used for eas build and update.
* When expo's metro config is used, you should fix it first if bundling fails.
    - Check why bundling fails.
    - For my case, some source code having `db` (`appBanner.db.js`) seems to be considered as assets file, so that finally bundling fails.
* After fixing bundling error, finally all (bundling, eas build and update) work.

## Detailed my case

### [0] Version

I set the latest expo compatible version.

```
    "expo": "~51.0.28",
    "react-native": "0.74.5",
```

### [0] When expo's metro config is used, the bundling fails.

use expo's metro config

```
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const config = {
  transformer: {
    assetPlugins: ['expo-asset/tools/hashAssetFiles'],
    babelTransformerPath: require.resolve('react-native-svg-transformer/react-native'),
  },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== 'svg'),
    sourceExts: [...sourceExts, 'svg'],
  },
};

module.exports = mergeConfig(defaultConfig, config);
```

It fails and says `unable to resolve module` ... mostly `XXX.db`

```
iOS ./index.js ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ 100.0% (3487/3487)[radon] Unable to resolve module ./app-banner.db from /node_modules/...SOMEWHERE_IN_MY_CODE.../index.js:
```

### [1] When react-native's metro config is used, bundling works, but eas build or update fails.

I tried to use react-native's metro config, but finally couldn't succeed in eas build and update.

### [2] Debugging

After a lot of building and testing, I checked the differences between expo and react native
I just console logged the config of expo and react native.

#### react-native metro config : bundling works

```
{
  "resolver": {
    "assetExts": [
      "bmp",
      "gif",
      "jpg",
      "jpeg",
      "png",
      "psd",
      "svg",
      "webp",
      "m4v",
      "mov",
      "mp4",
      "mpeg",
      "mpg",
      "webm",
      "aac",
      "aiff",
      "caf",
      "m4a",
      "mp3",
      "wav",
      "html",
      "pdf",
      "yaml",
      "yml",
      "otf",
      "ttf",
      "zip"
    ],
```

#### expo's metro config : bundling fails

`db` file extension is included in `assertExts`. It means that my source code like `appBanner.db.js` is recognized as assets, but source code.

```
{
  "resolver": {
    "assetExts": [
      "bmp",
      "gif",
      "jpg",
      "jpeg",
      "png",
      "psd",
      "svg",
      "webp",
      "m4v",
      "mov",
      "mp4",
      "mpeg",
      "mpg",
      "webm",
      "aac",
      "aiff",
      "caf",
      "m4a",
      "mp3",
      "wav",
      "html",
      "pdf",
      "yaml",
      "yml",
      "otf",
      "ttf",
      "zip",
      "heic",
      "avif",
      "db"      // 'db' is here !!!
    ],
```

### [3] Modification

not include `db` file in assets

```
const { getDefaultConfig } = require('expo/metro-config');

module.exports = (() => {
  const config = getDefaultConfig(__dirname);

  const { transformer, resolver } = config;

  config.transformer = {
    ...transformer,
    babelTransformerPath: require.resolve('react-native-svg-transformer/expo'),
  };
  config.resolver = {
    ...resolver,
    assetExts: resolver.assetExts
      .filter((ext) => ext !== 'svg')
      .filter((ext) => ext !== 'db'),                 // add to filter db file in assets
    sourceExts: [...resolver.sourceExts, 'svg'],
  };

  return config;
})();
```

### [4] It works

With this change, it works.

* bundling
* eas build
* eas update

**Comment by rizwan-gill** [comment]
After fixing this error getting this issue.
The development server returned the error code 404
https://github.com/expo/expo/issues/33602
Any solution how to fix this error's

**Comment by brentvatne** [comment]
@KrisLau - can you create a new issue with a minimal reproducible example please?

**Comment by KrisLau** [comment]
@brentvatne I finally figured it out sorry I feel like nuking myself instead of my android directory 💀 

When I upgraded react native my app/build.gradle originally had:

```gradle
react = [
   // Blah blah
]
```

At some point in the new version I noticed that it had changed to curly braces so I swapped to that but I didn't remove the `=` so I ended up with 

```gradle
react = {
   // Blah blah
}
```

I just noticed it's supposed to be 

```gradle
react {
   // Blah blah
}
``` 

Super dumb mistake haha and cost me more than a few days 💀

**Comment by aquinq** [comment]
For those using Sentry, with possibly a custom serializer, make sure to replace the default expo config by the one exported by Sentry https://docs.expo.dev/guides/using-sentry/#update-metro-configuration
```ts
// This replaces `const { getDefaultConfig } = require('expo/metro-config');`
const { getSentryExpoConfig } = require('@sentry/react-native/metro');

// This replaces `const config = getDefaultConfig(__dirname);`
const config = getSentryExpoConfig(__dirname);

module.exports = config;

```
